tools/cli_commands/erv2.py

```diff
@@ -0,0 +1,402 @@
+                        self.image,
+                    ],
+                    check=True,
+                    capture_output=True,
+                )
+
+            # # get the cdk.tf.json
+            with task(self.progress_spinner, "-- Copying the generated cdk.tf.json"):
+                run(
+                    [
+                        "docker",
+                        "cp",
+                        "erv2:/home/app/cdktf.out/stacks/CDKTF/cdk.tf.json",
+                        str(self.temp),
+                    ],
+                    check=True,
+                    capture_output=True,
+                )
+        except CalledProcessError as e:
+            print(e.stderr.decode("utf-8"))
+            print(e.stdout.decode("utf-8"))
+            raise
+
+
+class TfRun(Protocol):
+    def __call__(self, path: Path, cmd: list[str]) -> str: ...
+
+
+def tf_run(path: Path, cmd: list[str]) -> str:
+    env = os.environ.copy()
+    env["TF_CLI_ARGS"] = "-no-color"
+    try:
+        return run(
+            ["terraform", *cmd],
+            cwd=path,
+            check=True,
+            capture_output=True,
+            env=env,
+        ).stdout.decode("utf-8")
+    except CalledProcessError as e:
+        print(e.stderr.decode("utf-8"))
+        print(e.stdout.decode("utf-8"))
+        raise
+
+
+class TfAction(Enum):
+    CREATE = "create"
+    DESTROY = "delete"
+
+
+class TfResource(BaseModel):
+    address: str
+
+    @property
+    def id(self) -> str:
+        return self.address.split(".")[1]
+
+    @property
+    def type(self) -> str:
+        return self.address.split(".")[0]
+
+    def __str__(self) -> str:
+        return self.address
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def __lt__(self, other: TfResource) -> bool:
+        return self.address < other.address
+
+
+class TfResourceList(BaseModel):
+    resources: list[TfResource]
+
+    def __iter__(self) -> Iterator[TfResource]:  # type: ignore
+        return iter(self.resources)
+
+    def _get_resource_by_type(self, type: str) -> list[TfResource]:
+        results = [resource for resource in self.resources if resource.type == type]
+        if not results:
+            raise KeyError(f"Resource type {type} not found!")
+        return results
+
+    def __getitem__(self, tf_resource: TfResource) -> TfResource:
+        results = self._get_resource_by_type(tf_resource.type)
+        if len(results) == 1:
+            # best case scenario
+            return results[0]
+        # try to find the resource by fuzzy matching the id
+        # reverse sort to get the longest match first
+        for resource in sorted(results, reverse=True):
+            if tf_resource.id.startswith(resource.id):
+                # new resource id has a prefix
+                return resource
+            if tf_resource.id.endswith(resource.id):
+                # new resource id has a suffix
+                return resource
```

Comment:
This is a little confusing, Maybe some doc and examples in docstring?

Reply:
Added more comments in [7717f80](https://github.com/app-sre/qontract-reconcile/pull/4644/commits/7717f801a44905962a308ff10a3e6ef53b5802d5)

Reply:
what's the use case of

```python
        # reverse sort to get the longest match first
        for resource in sorted(results, reverse=True):
```

I tested it with `for resource in results`, tests still pass.

My understanding is the matching logic order is:

* address
* single resource within the type
* id match with optional prefix/suffix

If we want more fuzzy match, can use [difflib.get_close_matches](https://docs.python.org/3/library/difflib.html#difflib.get_close_matches).

Reply:
I've added a `reverse=True` test and some more docs in the last commit

Reply:
it works for suffix but not prefix, for example

I have these 2 resources

```python
            TfResource(
                address="aws_secretsmanager_secret.playground-user"
            ),
            TfResource(
                address="aws_secretsmanager_secret.foobar-playground-user"
            ),
```

then this test case will fail as it will match `playground-user` first.

```python
    assert (
        terraform_resource_list[
            TfResource(
                address="aws_secretsmanager_secret.secret-foobar-playground-user"
            )
        ].address
        == "aws_secretsmanager_secret.foobar-playground-user"
    )
```

Reply:
This is driving me crazy. I refactored the `TfResourceList.__getitem__` method, and now it returns a list of possible resources. The user must decide which one is the correct one.

Reply:
nice! Now

```python
    def __lt__(self, other: TfResource) -> bool:
        return self.address < other.address
```

can be removed since no sorting needed anymore.

Reply:
done